python/hsfs/core/vector_server.py

```diff
@@ -0,0 +1,417 @@
+                "please reinitialise prepared statements with `training_dataset.init_prepared_statement()` "
+                "or `feature_view.init_serving()`"
+            )
+
+        # create dict object that will have of order of the vector as key and values as
+        # vector itself to stitch them correctly if there are multiple feature groups involved. At this point we
+        # expect that backend will return correctly ordered vectors.
+        batch_dicts = {}
+
+        # check if primary key map correspond to serving_keys.
+        if not entry.keys() == self.serving_keys:
+            raise ValueError(
+                "Provided primary key map doesn't correspond to serving_keys"
+            )
+
+        # get schemas for complex features once
+        complex_features = self.get_complex_feature_schemas(vector_server)
+
+        self.refresh_mysql_connection(external)
+        for prepared_statement_index in self.prepared_statements:
+            order_in_batch = 0
+            prepared_statement = self.prepared_statements[
+                prepared_statement_index]
+            if preview_sample:
+                prepared_statement = self._make_preview_statement(
+                    prepared_statement, preview_sample)
+            with self.prepared_statement_engine.connect() as mysql_conn:
+                result_proxy = mysql_conn.execute(
+                    prepared_statement,
+                    batch_ids=None if preview_sample else tuple(
+                        zip(
+                            *[
+                                entry.get(key)
+                                for key in self._pkname_by_serving_index[
+                                    prepared_statement_index
+                                ]
+                            ]
+                        )
+                    ),
+                ).fetchall()
+            for row in result_proxy:
+                result_dict = self.deserialize_complex_features(
+                    complex_features, dict(row.items())
+                )
+                if not result_dict:
+                    raise Exception(
+                        "No data was retrieved from online feature store using input "
+                        + entry
+                    )
+                # apply transformation functions
+                result_dict = self._apply_transformation(
+                    vector_server.transformation_functions, result_dict
+                )
+                if order_in_batch in batch_dicts:
+                    batch_dicts[order_in_batch] += list(result_dict.values())
+                else:
+                    batch_dicts[order_in_batch] = list(result_dict.values())
+                order_in_batch += 1
+
+        if not len({len(i) for i in list(batch_dicts.values())}) == 1:
+            raise Exception(
+                "Not all feature vectors have the same length. Please make sure that all primary key values "
```

Comment:
I think there are two cases, assume we have two primary key columns pk1 and pk2:
1. user provides different number of primary keys for each: `{"pk1": [1,2], "pk2": [1,2,3]}`
2. user provides same number of primary keys for each: `{"pk1": [1,2,3], "pk2": [1,2,3]}` but it cannot find one or more of the pk values in the tables. -> This is the case that's checked here,

Till, you are right, we could also check for case 1, so far we just assumed it will then become case 2 by itself and fail there.

Reply:
I see, didn't think of case 2, indeed! 👍 I still think it makes sense to check for case 1 beforehand, because it's an easy catch and will contribute to a more responsive client.

Reply:
Agree, will check case 1